src/lib/reasoners/bitv_rel.ml

```diff
@@ -424,74 +642,284 @@ module Any_constraint = struct
+        (Z.extract (Z.lognot sup) 0 numbits_inf)
+        (Z.extract (Z.lognot inf) 0 numbits_inf)
+  else
+    (* Top [sz - numbits_sup] are 0 in both, the next significant bit differs *)
+    sz - numbits_sup
+
+(* If m and M are the minimal and maximal values of an union of intervals, the
+   longest sequence of most significant bits shared between m and M can be fixed
+   in the bit-vector domain; see "Is to BVs" in section 4.1 of
+
+   Sharpening Constraint Programming approaches for Bit-Vector Theory.
+   Zakaria Chihani, Bruno Marre, François Bobot, Sébastien Bardin.
+   CPAIOR 2017. International Conference on AI and OR Techniques in
+   Constraint Programming for Combinatorial Optimization Problems, Jun
+   2017, Padova, Italy.
+   https://cea.hal.science/cea-01795779/document *)
+let constrain_bitlist_from_interval bv int =
+  let open Domains.Ephemeral in
+  let inf, inf_ex, inf_large = Intervals.borne_inf int in
+  assert inf_large;
+  let inf = Q.to_bigint inf in
+  let sup, sup_ex, sup_large = Intervals.borne_sup int in
+  assert sup_large;
+  let sup = Q.to_bigint sup in
+
+  let sz = Bitlist.width !!bv in
+  let nshared = shared_msb sz inf sup in
+  if nshared > 0 then
+    let ex = Ex.union inf_ex sup_ex in
+    let shared_bl =
+      Bitlist.exact nshared (Z.extract inf (sz - nshared) nshared) ex
+    in
+    update ~ex bv @@
+    Bitlist.concat shared_bl (Bitlist.unknown (sz - nshared) Ex.empty)
+
+(* These are helpers for readability because [Intervals.mk_closed] makes my
+   brain melt *)
+
+type inequality = Large | Strict
+
+let is_large = function Large -> true | Strict -> false
+
+type bound = Infinite | Bounded of Q.t * inequality
+
+let large n = Bounded (n, Large)
+
+let strict n = Bounded (n, Strict)
+
+let strict_z n = strict (Q.of_bigint n)
+
+let infinite = Infinite
+
+let interval ~ex lb ub =
+  match lb, ub with
+  | Infinite, Infinite -> Intervals.undefined Tint
+  | Infinite, Bounded (ub, ineq) ->
+    Intervals.new_borne_sup ex ub ~is_le:(is_large ineq)
+      (Intervals.undefined Tint)
+  | Bounded (lb, ineq), Infinite ->
+    Intervals.new_borne_inf ex lb ~is_le:(is_large ineq)
+      (Intervals.undefined Tint)
+  | Bounded (lb, lineq), Bounded (ub, uineq) ->
+    Intervals.mk_closed
+      lb ub
+      (is_large lineq) (is_large uineq)
+      ex ex Tint
+
+(* Algorithm 1 from "Sharpening Constraint Programming approaches for
+   Bit-Vector Theory". *)
+let constrain_interval_from_bitlist int bv =
+  let open Interval_domains.Ephemeral in
+  let ex = Bitlist.explanation bv in
+  (* [exclude i1 i2] being [i2 \ i1] also makes my brain hurt *)
+  let remove i2 i1 = Intervals.exclude i1 i2 in
+  update ~ex int @@
+  List.fold_left (fun acc ((lb, _ex_lb), (ub, _ex_ub)) ->
+      (* Note: the bounds adjustment is expressed by removing specific
+         intervals that are forbidden by the bitlist domain. This is always
+         correct independently of the bounds' explanations: we only use the
```

Comment:
What is happening is:

 - We have a bitlist domain that has an explanation [ex].
 - We build some intervals that the bitlist domain proves are impossible (for instance, if the bitlist domain is `00??`, then the interval `[4, 15]` is impossible because no value in the interval `[4, 15]` starts with `00`).
 - We are removing those intervals from the existing interval domain using `exclude`

The intervals that we build from the bitlist domain only need the explanation from the bitlist domain to be valid, so we don't need to incorporate the explanations from the old bounds into it. The `exclude` and `intersect` functions from the intervals module takes care of appropriately (hopefully… but if not it is a bug in `exclude`, not in this code) updating the bounds to take into accound both explanations.

Is that more clear?

Reply:
This comment is no longer needed now that the iteration on the union-of-interval (correctly) do not expose the bounds after #1108